contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,148 @@
+pub extern crate r2d2;
+
+use std::collections::BTreeMap;
+use std::io;
+use std::marker::{Send, Sized};
+use rocket::config::{self, ConfigError, Value};
+
+use self::r2d2::ManageConnection;
+
+#[cfg(feature = "diesel")]
+pub extern crate diesel;
+
+#[cfg(feature = "pg_pool")]
+pub extern crate postgres;
+#[cfg(feature = "pg_pool")]
+pub extern crate r2d2_postgres;
+
+#[cfg(feature = "redis_pool")]
+pub extern crate redis;
+#[cfg(feature = "redis_pool")]
+pub extern crate r2d2_redis;
+
+/// A struct containing database configuration options.
+///
+/// This retrieves the key-value pairs specified in the `Rocket.toml` under
+/// the databases key for use in further configuring the desired connection.
+pub struct DatabaseConfig<'a> {
+    pub url: &'a str,
+    pub pool_size: u32,
+    pub extras: Option<&'a BTreeMap<String, Value>>,
+}
+
+/// A custom error type to wrap multiple, disparate error types
+///
+/// This allows for returning the given error from the connection manager
+/// creation methods and the r2d2 connection pool methods.
+pub enum DbError<T> {
+    Custom(T),
+    PoolError(r2d2::Error),
+}
+
+/// This method retrieves the database configuration from the `Rocket.toml` file
+/// and returns a DatabaseConfig struct.
+pub fn database_config<'a>(name: &str, from: &'a config::Config) -> Option<DatabaseConfig<'a>> {
+    from.get_slice("databases").and_then(|db_config| {
+        db_config
+            .iter()
+            .filter(|element|
+                match element.as_table().and_then(|table|
+                    Some(table["name"].as_str() == Some(name))
+                ) {
+                    Some(true) => true,
+                    _ => false,
+                }
+            )
+            .nth(0)
+            .ok_or(ConfigError::UnknownKey(format!("Could not find key: {}", name)))
+    })
+    .and_then(|connection_config| {
+        Ok(connection_config["url"].as_str().and_then(|url| {
+            Some(DatabaseConfig {
+                url,
+                pool_size: connection_config["pool_size"].as_integer().unwrap_or(from.workers as i64) as u32,
+                extras: connection_config.as_table(),
+            })
+        }))
+    })
+    .ok()
+    .and_then(|result| result)
+}
+
+pub fn available_database_configs<'a>(config: &'a config::Config) -> Vec<&'a str> {
```

Comment:
What is this function for? I don't see it being used anywhere.

Reply:
This was incomplete but meant to log out available configurations for when the given database name can't be found while the procedural macro is being invoked. We can scrap this or proceed with it.